File: apps/desktop/src/commands/smart-feed.ts
import {
  SmartFeed,
  type SmartFeedAPI,
  SmartFeedCreate,
  SmartFeedList,
  SmartFeedUpdate,
} from '@colette/core'
import { invoke } from '@tauri-apps/api/core'

export class SmartFeedCommands implements SmartFeedAPI {
  list(): Promise<SmartFeedList> {
    return invoke('list_smart_feeds', {}).then(SmartFeedList.parse)
  }

  get(id: string): Promise<SmartFeed> {
    return invoke('get_smart_feed', { id }).then(SmartFeed.parse)
  }

  create(data: SmartFeedCreate): Promise<SmartFeed> {
    return invoke('create_smart_feed', {
      data: SmartFeedCreate.parse(data),
    }).then(SmartFeed.parse)
  }

  update(id: string, data: SmartFeedUpdate): Promise<SmartFeed> {
    return invoke('update_smart_feed', {
      id,
      data: SmartFeedUpdate.parse(data),
    }).then(SmartFeed.parse)
  }

  delete(id: string): Promise<void> {
    return invoke('delete_smart_feed', { id }).then()
  }
}

File: apps/desktop/src-tauri/src/command/smart_feed.rs
use colette_api::{
    smart_feed::{SmartFeed, SmartFeedCreate, SmartFeedUpdate},
    Paginated, Session,
};
use colette_core::smart_feed::SmartFeedService;
use tauri::State;
use uuid::Uuid;

#[tauri::command]
pub async fn list_smart_feeds(
    service: State<'_, SmartFeedService>,
    session: State<'_, Session>,
) -> Result<Paginated<SmartFeed>, String> {
    let smart_feeds = service
        .list_smart_feeds(session.profile_id)
        .await
        .map_err(|e| e.to_string())?;

    Ok(smart_feeds.into())
}

#[tauri::command]
pub async fn create_smart_feed(
    service: State<'_, SmartFeedService>,
    session: State<'_, Session>,
    data: SmartFeedCreate,
) -> Result<SmartFeed, String> {
    let smart_feed = service
        .create_smart_feed(data.into(), session.profile_id)
        .await
        .map_err(|e| e.to_string())?;

    Ok(smart_feed.into())
}

#[tauri::command]
pub async fn get_smart_feed(
    service: State<'_, SmartFeedService>,
    session: State<'_, Session>,
    id: Uuid,
) -> Result<SmartFeed, String> {
    let smart_feed = service
        .get_smart_feed(id, session.profile_id)
        .await
        .map_err(|e| e.to_string())?;

    Ok(smart_feed.into())
}

#[tauri::command]
pub async fn update_smart_feed(
    service: State<'_, SmartFeedService>,
    session: State<'_, Session>,
    id: Uuid,
    data: SmartFeedUpdate,
) -> Result<SmartFeed, String> {
    let smart_feed = service
        .update_smart_feed(id, data.into(), session.profile_id)
        .await
        .map_err(|e| e.to_string())?;

    Ok(smart_feed.into())
}

#[tauri::command]
pub async fn delete_smart_feed(
    service: State<'_, SmartFeedService>,
    session: State<'_, Session>,
    id: Uuid,
) -> Result<(), String> {
    service
        .delete_smart_feed(id, session.profile_id)
        .await
        .map_err(|e| e.to_string())
}
